src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/OpenLdapUnsafeMethods.cs

```diff
@@ -0,0 +1,192 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+
+namespace System.DirectoryServices.Protocols
+{
+    internal class OpenLDAP
+    {
+        /// <summary>
+        /// We need to specify the versioned full-name library in order to not require a native shim to
+        /// P/Invoke into the library.
+        /// </summary>
+        private const string _openldapdll = "libldap-2.4.so.2";
```

Comment:
FYI: In order to not require a Native shim (since it is not practical as this assembly ships OOB) we have to specify the full name of the so when we PInvoke. This library is already installed (with this version) on most distros already, and version hasn't changed since 2007 which is why we chose to PInvoke directly.

Reply:
what would be the behavior when this not installed?

Also, would be confirm with the runtime team if it is ok to have pinvokes like these while the library can be absent from the system. I am asking this because in .NET Native the app used to throw while loading. I hope we don't have such cases any more.


---
In reply to: [414199460](https://github.com/dotnet/runtime/pull/35380#discussion_r414199460) [](ancestors = 414199460)

Reply:
You would get an AssemblyNotFound Exception, but there are very few distros where that would happen.